dvc/repo/ls.py

```diff
@@ -0,0 +1,108 @@
+import os
+
+from dvc.exceptions import DvcException, NotDvcRepoError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outputs_only=False):
+    if os.path.exists(url):
+        if rev:
+            raise DvcException("Local repo doesn't support revisions")
+        return _ls_local_repo(url, target, recursive, outputs_only)
+
+    return _ls_remote_git_repo(url, target, rev, recursive, outputs_only)
+
+
+def _ls_files_local_repo(path, target=None, recursive=None):
+    from dvc.scm.tree import WorkingTree
+    from dvc.ignore import CleanTree
+
+    def prettify(prefix, filename):
+        if not prefix == os.curdir:
+            filename = os.path.join(prefix, filename)
+        return filename
+
+    files = []
+    target_path = os.path.join(path, target) if target else path
+    tree = CleanTree(WorkingTree(target_path))
+
+    try:
+        target_path = os.path.join(path, target) if target else path
+        if not os.path.exists(target_path):
+            raise FileNotFoundError("{} doesn't exist".format(target_path))
+
+        if target:
+            files.append(prettify(os.curdir, target))
+            if not recursive:
+                return files
+
+        for dirpath, dirnames, filenames in tree.walk(target_path):
+            prefix = os.path.relpath(dirpath, path)
+            files.extend(map(lambda f: prettify(prefix, f), dirnames))
+            files.extend(map(lambda f: prettify(prefix, f), filenames))
+            if not recursive:
+                break
+    except (NotADirectoryError, FileNotFoundError):
```

Comment:
So if you run list with non-existing target
```
$ dvc list https://github.com/iterative/dvc.org sdfsdf                                                                                      
```
it won't show anything (except for a newline) and will finish successfully. It doesn't seem right to me.

Reply:
> it won't show anything (except for a newline) and will finish successfully. It doesn't seem right to me

What do you expect to see, an error? I think empty output is correct behavior in such case cause target is applied as a filter, so passing wrong/doesn't existed target means [] files output

I will add test for the case

Reply:
Well, for now it is not really a filter but rather a recursive list, right? It accepts something that should exist (e.g. directory) and lists everything under it. So this specific behaviour is similar to `$ ls non-existing` which raises an error and further away from `aws s3 ls` which always treats your argument as a filter.

Reply:
> Well, for now it is not really a filter but rather a recursive list, right? It accepts something that should exist (e.g. directory) and lists everything under it

In fact no, generally based on the https://github.com/iterative/dvc/pull/3246#discussion_r373874769 `dvc list url target` checks does the target exists or not. Only if we provide `-R` flag it becomes recursive and shows content

> this specific behaviour is similar to $ ls non-existing which raises an error

I think this behavior is covered with `dvc list not_existed` which throws an error (I think the error message should be improved)
```
(dvc) jiojiajiu@pon4ik:~$ dvc list not_existed
ERROR: failed to list 'not_existed' - Failed to clone repo 'not_existed' to '/tmp/tmp7pu3j_76dvc-clone': Cmd('git') failed due to: exit code(128)
  cmdline: git clone --no-single-branch -v not_existed /tmp/tmp7pu3j_76dvc-clone
  stderr: 'fatal: repository 'not_existed' does not exist
```

but with `dvc list repo target` target works as a filter
```
dvc list repo not_exitsted

```
and returns nothing if there is no anything that matches

I think cause we always show the target if it matches, pointing empty output means that nothing was found with such target
What do you think @efiop

Reply:
@JIoJIaJIu I think we should raise an error after all. Additional reason is that `dvc get/import` in that case would raise an error.